web.go

```diff
@@ -0,0 +1,190 @@
+
+import (
+	"context"
+	"github.com/gorilla/websocket"
+	"github.com/skysoft-atm/gorillaz/mux"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"time"
+)
+
+// serves a file as an http response
+func ServeFileFunc(file string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "text/html; charset=UTF-8")
+		b, err := ioutil.ReadFile(filepath.Clean(file))
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+		} else {
+			w.WriteHeader(http.StatusOK)
+			_, err = w.Write(b)
+			if err != nil {
+				Log.Debug("Could not write response on http connection", zap.Error(err))
+			}
+		}
+	}
+}
+
+type WebsocketConfig struct {
+	WriteWait  time.Duration // Time allowed to write a message to the peer.
+	PongWait   time.Duration // Time allowed to read the next pong message from the peer.
+	PingPeriod time.Duration // Send pings to peer with this period. Must be less than pongWait.
+}
+
+type WebsocketOption func(*WebsocketConfig)
+
+// Websocket message
+type WsMessage struct {
+	MessageType int
+	Data        []byte
+}
+
+// Upgrades the http request to a websocket connection
+func UpgradeToWebsocketWithContext(rw http.ResponseWriter, req *http.Request, opts ...WebsocketOption) (chan<- *WsMessage, context.Context, error) {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+	c := WebsocketConfig{
+		WriteWait:  10 * time.Second,
+		PongWait:   60 * time.Second,
+		PingPeriod: (60 * time.Second * 9) / 10,
+	}
+	for _, o := range opts {
+		o(&c)
+	}
+	messageChan := make(chan *WsMessage, 1000)
+	conn, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = conn.SetWriteDeadline(time.Now().Add(c.WriteWait))
+	if err != nil {
+		return nil, nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		ticker := time.NewTicker(c.PingPeriod)
+		defer func() {
+			ticker.Stop()
+			err := conn.Close()
+			if err != nil {
+				Log.Debug("Could not close websocket connection", zap.Error(err))
+			}
+			// empty the ticker channel
+			select {
+			case <-ticker.C:
+			default:
+			}
+		}()
+		for {
+			select {
+			case msg := <-messageChan:
+				err := conn.SetWriteDeadline(time.Now().Add(c.WriteWait))
+				if err != nil {
+					Log.Debug("Could not set write deadline on websocket connection", zap.Error(err))
+				}
+				if err := conn.WriteMessage(msg.MessageType, msg.Data); err != nil {
+					Log.Debug("Could not write message over websocket", zap.Error(err))
+					cancel()
+					return
+				}
+			case <-ticker.C:
+				err := conn.SetWriteDeadline(time.Now().Add(c.WriteWait))
+				if err != nil {
+					Log.Debug("Could not set write deadline on websocket connection", zap.Error(err))
+				}
+				if err := conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+					Log.Debug("Could not send ping message over websocket", zap.Error(err))
```

Comment:
Log in error here?

Reply:
this would also log an error when the client closes the page in the browser